File: 62-unique-paths/62-unique-paths.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        
        def memoize(i, j, dp):
            if dp[i][j] != -1:
                return dp[i][j]
            
            if i == 0 and j == 0:
                return 1
            
            dp[i][j] = 0
            
            if i == 0: dp[i][j] = memoize(i, j - 1, dp)
            elif j == 0: dp[i][j] = memoize(i - 1, j, dp)
            else:
                dp[i][j] = memoize(i, j - 1, dp) + memoize(i - 1, j, dp)
            
            return dp[i][j]
            
        dp = [[-1] * n for _ in range(m)]
        return memoize(m - 1, n - 1, dp)

File: 62-unique-paths/62-unique-paths.cpp
class Solution {
public:
    int uniquePaths(int m, int n) {
        // vector<vector<int>> dp(m + 1, vector<int>(n + 1, -1));
        // return aakashSir(m, n, 0, 0, dp);
        
        return aakashSirTab(m, n);
    }
private:
    int aakashSir(int m, int n, int i, int j, vector<vector<int>>& dp) {
        if(i >= m && j >= n) return 0;
        
        if(i == m - 1 && j == n - 1)
            return 1;
        
        if(dp[i][j] != -1) return dp[i][j];
        
        if(i >= m) return dp[i][j] = aakashSir(m, n, i, j + 1, dp);
        if(j >= n) return dp[i][j] = aakashSir(m, n, i + 1, j, dp);
        
        return dp[i][j] = aakashSir(m, n, i, j + 1, dp) + aakashSir(m, n, i + 1, j, dp);
    }
    
    int aakashSirTab(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n, 0));
        
        for(int i = 0; i < m; ++i)
            dp[i][0] = 1;
        
        for(int i = 0; i < n; ++i)
            dp[0][i] = 1;
        
        for(int i = 1; i < m; ++i) {
            for(int j = 1; j < n; ++j) {
                dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        
        return dp[m - 1][n - 1];
    }
};